File: client/js/hero.js
define(['character',
  'player',
  'mob',
  'exceptions',
  'inventory',
  'skillbar',
  'chest',
  'npc'
], function (
  Character,
  Player,
  Mob,
  Exceptions,
  Inventory,
  Skillbar,
  Chest,
  Npc) {

  var Hero = Player.extend({
    init: function (id, name) {
      this._super(id, name, Types.Entities.PLAYER);

      // storage
      this.storage = null;

      this.inventory = new Inventory();
      this.skillbar = new Skillbar();
    },

    isHostile: function (entity) {
      return (
             (entity instanceof Mob) 
             || ((entity instanceof Player) 
                 && this.id != entity.id
                 && (!this.guild 
                     || !entity.guild 
                     || this.guild.name != entity.guild.name)
                )
      );
    },

    die: function () {
      this.removeTarget();
      this.isDead = true;

      log.info(this.id + " is dead");

      this.isDying = true;

      this.stopBlinking();
      this.setSprite(globalSprites["death"]);

      this.animate("death", 120, 1, function () {
        log.info(this.id + " was removed");

        setTimeout(function () {
          globalGame.removeEntity(this);
          globalGame.removeFromRenderingGrid(this, this.gridX, this.gridY);

          globalGame.audioManager.fadeOutCurrentMusic();
          globalGame.audioManager.playSound("death");

          globalGame.entities = {};
          globalGame.deathpositions = {};
          globalGame.currentCursor = null;
          globalGame.zoningQueue = [];
          globalGame.previousClickPosition = {};

          globalGame.initPathingGrid();
          globalGame.initEntityGrid();
          globalGame.initRenderingGrid();
          globalGame.initItemGrid();

          globalGame.selectedX = 0;
          globalGame.selectedY = 0;
          globalGame.selectedCellVisible = false;
          globalGame.targetColor = "rgba(255, 255, 255, 0.5)";
          globalGame.targetCellVisible = true;
          globalGame.hoveringTarget = false;
          globalGame.hoveringPlayer = false;
          globalGame.hoveringMob = false;
          globalGame.hoveringItem = false;
          globalGame.hoveringCollidingTile = false;

          globalGame.playerDeath();
        }.bind(this), 1000);
      }.bind(this));

      this.forEachAttacker(function (attacker) {
        attacker.disengage();
        attacker.idle();
      });
    },

    checkAggro: function () {
      globalGame.forEachMob(function (mob) {
        if (mob.isAggressive && !mob.isAttacking() && this.isNear(mob, mob.aggroRange)) {
          this.aggro(mob);
        }
      }.bind(this));
    },

    aggro: function (character) {
      if (!character.isWaitingToAttack(this) && !this.isAttackedBy(character)) {
        this.log_info("Aggroed by " + character.id + " at (" + this.gridX + ", " + this.gridY + ")");
        globalGame.client.sendAggro(character);
        character.waitToAttack(this);
      }
    },

    beforeStep: function () {
      var blockingEntity = globalGame.getEntityAt(this.nextGridX, this.nextGridY);
      if (blockingEntity && blockingEntity.id !== this.id) {
        //log.debug("Blocked by " + blockingEntity.id);
      }

      globalGame.unregisterEntityPosition(this);
    },

    doStep: function () {
      if (this.hasNextStep()) {
        globalGame.registerEntityDualPosition(this);
      }

      if (globalGame.isZoningTile(this.gridX, this.gridY)) {
        globalGame.enqueueZoningFrom(this.gridX, this.gridY);
      }

      this.forEachAttacker(function (attacker) {
        if (attacker.isAdjacent(attacker.target)) {
          attacker.lookAtTarget();
        } else {
          attacker.follow(this);
        }
      }.bind(this));

      if ((this.gridX <= 85 && this.gridY <= 179 && this.gridY > 178) ||  (this.gridX <= 85 && this.gridY <= 266 && this.gridY > 265)) {
        globalGame.tryUnlockingAchievement("INTO_THE_WILD");
      }

      if (this.gridX <= 85 && this.gridY <= 293 && this.gridY > 292) {
        globalGame.tryUnlockingAchievement("AT_WORLDS_END");
      }

      if (this.gridX <= 85 && this.gridY <= 100 && this.gridY > 99) {
        globalGame.tryUnlockingAchievement("NO_MANS_LAND");
      }

      if (this.gridX <= 85 && this.gridY <= 51 && this.gridY > 50) {
        globalGame.tryUnlockingAchievement("HOT_SPOT");
      }

      if (this.gridX <= 27 && this.gridY <= 123 && this.gridY > 112) {
        globalGame.tryUnlockingAchievement("TOMB_RAIDER");
      }

      globalGame.updatePlayerCheckpoint();

      if (!this.isDead) {
        globalGame.audioManager.updateMusic();
      }
    },

    startPathing: function (path) {
      var i = path.length - 1,
        x = path[i][0],
        y = path[i][1];

      if (this.isMovingToLoot()) {
        this.isLootMoving = false;
      } else if (!this.isAttacking()) {
        globalGame.client.sendMove(x, y);
      }

      // Target cursor position
      globalGame.selectedX = x;
      globalGame.selectedY = y;

      if (globalGame.renderer.mobile || globalGame.renderer.tablet) {
        globalGame.drawTarget = true;
        globalGame.clearTarget = true;
        globalGame.renderer.targetRect = globalGame.renderer.getTargetBoundingRect();
        globalGame.checkOtherDirtyRects(globalGame.renderer.targetRect, null, globalGame.selectedX, globalGame.selectedY);
      }
    },
    stopPathing: function (x, y) {
      globalGame.selectedCellVisible = false;

      if (globalGame.isItemAt(x, y)) {
        var item = globalGame.getItemAt(x, y);

        // notify the server that the user is trying
        // to loot the item
        globalGame.client.sendLoot(item);
      }

      if (!this.hasTarget() && globalGame.map.isDoor(x, y)) {
        var dest = globalGame.map.getDoorDestination(x, y);
        globalGame.teleport(dest);
      }

      if (this.target instanceof Npc) {
        globalGame.makeNpcTalk(this.target);
      } else if (this.target instanceof Chest) {
        globalGame.client.sendOpen(this.target);
        globalGame.audioManager.playSound("chest");
      }

      this.forEachAttacker(function (attacker) {
        if (!attacker.isAdjacentNonDiagonal(this)) {
          attacker.follow(this);
        }
      }.bind(this));

      globalGame.unregisterEntityPosition(this);
      globalGame.registerEntityPosition(this);
    },

    get areaName() {
      if (globalGame.audioManager.getSurroundingMusic(this)) {
        return globalGame.audioManager.getSurroundingMusic(this).name;
      }
    },

    loadInventory: function (data) {
      if (this.inventory) {
        this.inventory.loadFromObject(data);
      } else {
        this.inventory = new Inventory(data);
      }
    },

    loadSkillbar: function (data) {
      if (this.skillbar) {
        this.skillbar.loadFromObject(data);
      } else {
        this.skillbar = new Skillbar(data);
      }
    },

    lootedArmor: function (item) {
      // make sure that it's better than what we already have, and if so - equip it
      if (Types.getArmorRank(item.kind) > Types.getArmorRank(this.armor)) {
        item.use();

        // we are optimistically equipping the item before-hand.
        this.switchArmor(item);
      }
    },

    lootedWeapon: function (item) {
      // make sure that it's better than what we already have, and if so - equip it
      if (Types.getWeaponRank(item.kind) > Types.getWeaponRank(this.weapon)) {
        item.use();

        // we are optimistically equipping the item before-hand.
        this.switchWeapon(item);
      }
    },

    changedEquipment: function () {
      globalGame.storage.savePlayer(globalGame.renderer.getPlayerImage(),
        this);
      globalGame.playerChangedEquipment();
    },

    startInvincibility: function () {
      if (this.invincible) {
        // If the player already has invincibility, just reset its duration.
        if (this.invincibleTimeout) {
          clearTimeout(this.invincibleTimeout);
        }
      } else {
        this.invincible = true;
        globalGame.playerInvincible(true);
      }

      this.invincibleTimeout = setTimeout(function () {
        this.stopInvincibility();
        this.idle();
      }.bind(this), 15000);
    },

    stopInvincibility: function () {
      globalGame.playerInvincible(false);
      this.invincible = false;

      if (this.invincibleTimeout) {
        clearTimeout(this.invincibleTimeout);
      }
    },

    equip: function (itemKind) {
      this._super(itemKind);
      globalGame.app.initEquipmentIcons();
    },

    setStorage: function (storage) {
      this.storage = storage;
    },

    loadFromStorage: function (callback) {
      return;

      if (this.storage && this.storage.hasAlreadyPlayed()) {
        this.armor = this.storage.data.player.armor;
        this.weapon = this.storage.data.player.weapon;
      }

      log.debug("Loaded from storage");

      if (callback) {
        callback();
      }
    },

    updateStorage: function (callback) {
      return;

      if (this.storage) {
        this.storage.data.player.name = this.name;
        this.storage.data.player.armor = this.armor;
        this.storage.data.player.weapon = this.weapon;
      }

      log.debug("Updated storage");

      if (callback) {
        callback();
      }
    },

    loadFromObject: function (data) {
      // x and y in server are mapped to gridX and gridY on client
      this.setGridPosition(data.x, data.y);
      delete data.x;
      delete data.y;

      // set inventory data
      this.loadInventory(data.inventory);
      delete data.inventory;

      // set skillbar data
      this.loadSkillbar(data.skillbar);
      delete data.skillbar;

      $.extend(this, data);
    }
  });

  return Hero;
});

File: client/dart/hero.dart
library hero;

import "achievements.dart";
import "character.dart";
import "chest.dart";
import "game.dart";
import "inventory.dart";
import "inventoryitem.dart";
import "item.dart";
import "npc.dart";
import "player.dart";
import "skillbar.dart";
import "localstorage.dart";
import "lib/gametypes.dart";
import 'position.dart';

class Hero extends Player {

  Set<Orientation> directions = new Set<Orientation>();

  Inventory inventory;
  Skillbar skillbar;
  LocalStorage storage;

  Hero(int id, String name): super(id, name, Entities.PLAYER) {
    this.skillbar = new Skillbar([]);
    this.inventory = new Inventory([]);
  }

  void die() {
    this.removeTarget();

    this.isDead = true;
    this.isDying = true;

    this.log_info("is dead");

    this.stopBlinking();

    this.animate("death", 120, 1, () {
      this.log_info("was removed");

      this.trigger("Death");
    });

    this.forEachAttacker((Character attacker) {
      attacker.disengage();
      attacker.idle();
    });
  }

  /**
   * This function does pre-step preparations - as for now, just unregister
   * the previous entity position on the grid.
   */
  void beforeStep() {
    Game.unregisterEntityPosition(this);
  }

  void doStep() {
    super.doStep();

    if (Game.isZoningTile(this.gridPosition)) {
      Game.enqueueZoningFrom(this.gridPosition);
    }

    if ((this.gridPosition.x <= 85 && this.gridPosition.y <= 179 && this.gridPosition.y > 178) || (this.gridPosition.x <= 85 && this.gridPosition.y <= 266 && this.gridPosition.y > 265)) {
      Game.tryUnlockingAchievement(Achievement.INTO_THE_WILD);
    }

    if (this.gridPosition.x <= 85 && this.gridPosition.y <= 293 && this.gridPosition.y > 292) {
      Game.tryUnlockingAchievement(Achievement.AT_WORLDS_END);
    }

    if (this.gridPosition.x <= 85 && this.gridPosition.y <= 100 && this.gridPosition.y > 99) {
      Game.tryUnlockingAchievement(Achievement.NO_MANS_LAND);
    }

    if (this.gridPosition.x <= 85 && this.gridPosition.y <= 51 && this.gridPosition.y > 50) {
      Game.tryUnlockingAchievement(Achievement.HOT_SPOT);
    }

    if (this.gridPosition.x <= 27 && this.gridPosition.y <= 123 && this.gridPosition.y > 112) {
      Game.tryUnlockingAchievement(Achievement.TOMB_RAIDER);
    }

    Game.updatePlayerCheckpoint();

    if (!this.isDead) {
      Game.audioManager.updateMusic();
    }
  }

  void startPathing(List<List<int>> path) {
    int i = path.length - 1;
    int x = path[i][0];
    int y = path[i][1];

    if (this.isLootMoving) {
      this.isLootMoving = false;
    } else if (!this.isAttacking()) {
      Game.client.sendMove(new Position(x, y));
    }

    // Target cursor position
    Game.selected = new Position(x, y);
  }
  void stopPathing(Position position) {
    Game.selectedCellVisible = false;

    if (Game.isItemAt(position)) {
      Item item = Game.getItemAt(position);

      // notify the server that the user is trying
      // to loot the item
      Game.client.sendLoot(item);
    }

    if (!this.hasTarget() && Game.map.isDoor(position)) {
      Game.teleport(Game.map.getDoorDestination(position));
    }

    if (this.target is Npc) {
      Game.makeNpcTalk(this.target as Npc);
    } else if (this.target is Chest) {
      Game.client.sendOpen(this.target as Chest);
      Game.audioManager.playSound("chest");
    }

    this.forEachAttacker((Character attacker) {
      if (!attacker.isAdjacentNonDiagonal(this)) {
        attacker.follow(this);
      }
    });

    Game.unregisterEntityPosition(this);
    Game.registerEntityPosition(this);
  }

  String get areaName {
    if (Game.audioManager.getSurroundingMusic(this) != null) {
      return Game.audioManager.getSurroundingMusic(this).name;
    }

    return super.areaName;
  }

  void loadInventory(data) {
    if (this.inventory == null) {
      this.inventory = new Inventory(data);
      return;
    }

    this.inventory.loadFromObject(data);
  }

  void loadSkillbar(data) {
    if (this.skillbar == null) {
      this.skillbar = new Skillbar(data);
      return;
    }

    this.skillbar.loadFromObject(data);
  }

  void lootedArmor(Item item) {
    // make sure that it's better than what we already have, and if so - equip it
    if (this.getArmorRank() >= Types.getArmorRank(item.kind)) {
      return;
    }

    if (item is InventoryItem) {
      item.use();
    }

    // we are optimistically equipping the item before-hand.
    this.switchArmor(item);
  }

  void lootedWeapon(Item item) {
    // make sure that it's better than what we already have, and if so - equip it
    if (this.getWeaponRank() >= Types.getWeaponRank(item.kind)) {
      return;
    }

    if (item is InventoryItem) {
      item.use();
    }

    // we are optimistically equipping the item before-hand.
    this.switchWeapon(item);
  }

  void startInvincibility() {
    if (!this.isInvincible) {
      Game.playerInvincible(true);
    }

    super.startInvincibility();
  }

  void stopInvincibility() {
    Game.playerInvincible(false);

    super.stopInvincibility();
  }

  void equip(EntityKind itemKind) {
    super.equip(itemKind);

    Game.app.initEquipmentIcons();
  }

  void updateStorage() {
    if (this.storage == null) {
      return;
    }

    this.storage.updatePlayer(this);
  }
  
  List<Character> getNearestEnemies() {
    List<Character> nearestEnemies = Game.characters
      .where((Character character) => this != character && this.isHostile(character) && !character.isDead)
      .toList();
    nearestEnemies.sort((Character a, Character b) => this.distanceTo(a) - this.distanceTo(b));
    return nearestEnemies;
  }

  void loadFromObject(data) {
    this.loadInventory(data['inventory']);
    data.remove("inventory");

    // TODO(skillbar): implement
    /*this.loadSkillbar(data.skillbar);*/
    /*data.remove("skillbar");*/

    // we need to load the id as we start with a fake id
    this.id = data['id'];

    super.loadFromObject(data);
  }
}
